### PR TITLE
fix(headless-execution): MainMenuState can handle headless now. Correct loadingstate for headless clients.

### PR DESCRIPTION
**Pull request by DarkWeird**
### Contains

* Support for headless client in LoadingState.
* MainMenuState handles headless too.

> MTE's client not working yet. There needs to revisit `@RegisterSystem#value` (RegisterMode) 
> engine don't support headless clients at `Systems` level.. needs to separate Systems which handles GUI and Systems which handles client logic.

### How to test

Run Terasology in:
1. SinglePlayer
2. Join Game 
3. Host Game
4. Dedicated server (`./gradlew server`)

### CODE REVIEW

File: engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java [hunk 1]
```diff
@@ -45,38 +47,46 @@ public StateMainMenu(String showMessageOnLoad) {
...
+        if (!headless) {
+            //guiManager.openWindow("main");
+            context.get(NUIManager.class).pushScreen("engine:mainMenuScreen");
+        }
         if (!messageOnLoad.isEmpty()) {
             TranslationSystem translationSystem = context.get(TranslationSystem.class);
-            MessagePopup popup = nuiManager.pushScreen(MessagePopup.ASSET_URI, MessagePopup.class);
-            popup.setMessage("Error", translationSystem.translate(messageOnLoad));
+            if (headless) {
+                throw new RuntimeException("Game receive error and fall to main menu. See logs before");
```
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java:85]: Let's add messageOnLoad to this exception. Yes, something else _probably_ logged something when it set this messageOnLoad field, but I often have trouble actually finding those messages.

File: engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java [hunk 2]
```diff
@@ -155,11 +168,13 @@ public String getLoggingPhase() {
 
     @Override
     public boolean isHibernationAllowed() {
-        return true;
+        return !headless;
```
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java:171]: ?

**Review by keturn — COMMENTED** [review]
Looks reasonable, but yeah, let's be sure to test before merging.

File: engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java (hunk 2 above)
Comment (DarkWeird) [reply engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java:171]: Headless mode is not needed in hibernation.

HibernationSystem lower frame rate  while game off-focus.

P.S. it is useless:
1. MainMenu don't receive any tick.
2. Seems MTE don't starts hibernationSubsystem.

File: engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java [hunk 3]
```diff
@@ -45,38 +47,50 @@ public StateMainMenu(String showMessageOnLoad) {
...
+            context.get(NUIManager.class).pushScreen("engine:mainMenuScreen");
+        }
         if (!messageOnLoad.isEmpty()) {
             TranslationSystem translationSystem = context.get(TranslationSystem.class);
-            MessagePopup popup = nuiManager.pushScreen(MessagePopup.ASSET_URI, MessagePopup.class);
-            popup.setMessage("Error", translationSystem.translate(messageOnLoad));
+            if (headless) {
+                throw new RuntimeException(
+                        String.format(
+                                "Game receive error and fall to main menu: [%s]. See logs before",
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/core/modes/StateMainMenu.java:87]:
```suggestion
                                "Game could not be started, server attempted to return to main menu: [%s]. See logs before",
```

**Review by jdrueckert — COMMENTED** [review]
Tests out fine in singleplayer and multiplayer (host, join) on both headed and headless

**Comment by keturn** [comment]
I ran some MTE tests. Can confirm that throwing the exception there makes the message stand out better in some tests.

I'll re-run some of these to figure out if their failures are related to this branch or not.

**Review by keturn — APPROVED** [review]
After re-running those MTE tests, it didn't seem like their problems were particular to this branch.

**Pull request merged**